lib/src/stream_extensions.dart

```diff
@@ -33,4 +33,24 @@ extension StreamExtensions<T> on Stream<T> {
       sink.close();
     }));
   }
+
+  /// Returns the first data or error event emitted by [this] if it emits any,
+  /// or `null` if it emits a done event first.
+  Future<T?> get firstOrNull {
+    var completer = Completer<T?>.sync();
+    final subscription = listen(null);
```

Comment:
Can use `cancelOnError: true`. Then you can just tear off `completer.completeError`instead of creating a new closure. It can even be:

```dart
  var subscription = listen(null, 
    onError: completer.completeError,
    onDone: () { 
      completer.complete(null);
    },
    cancelOnError: true);
  subscription.onData((event) {
    subscription.cancel();
    completer.complete(event);
  });
```

Reply:
I considered doing it that way, but I think the symmetry between the `onData` and `onError` events is easier to follow than having one of them manually cancel the subscription and the other cancel it through `cancelOnError`.